gst/nnstreamer/tensor_source/tensor_src_iio.c

```diff
@@ -873,6 +959,143 @@ gst_tensor_set_all_channels (GstTensorSrcIIO * self, const gint contents)
+gst_tensor_src_iio_get_type_for_channel (GstTensorSrcIIOChannelProperties *
+    channel_prop)
+{
+  tensor_type ret;
+
+  if (channel_prop->is_signed) {
+    switch (channel_prop->storage_bits) {
+      case 8:
+        ret = _NNS_INT8;
+        break;
+      case 16:
+        ret = _NNS_INT16;
+        break;
+      case 32:
+        ret = _NNS_INT32;
+        break;
+      case 64:
+        ret = _NNS_INT64;
+        break;
+      default:
+        ret = _NNS_END;
+        break;
+    }
+  } else {
+    switch (channel_prop->storage_bits) {
+      case 8:
+        ret = _NNS_UINT8;
+        break;
+      case 16:
+        ret = _NNS_UINT16;
+        break;
+      case 32:
+        ret = _NNS_UINT32;
+        break;
+      case 64:
+        ret = _NNS_UINT64;
+        break;
+      default:
+        ret = _NNS_END;
+        break;
+    }
+  }
+  return ret;
+}
+
+/**
+ * @brief create the structure for the caps to update the src pad caps
+ * @param[in/out] structure Caps structure which will filled
+ * @returns True if structure is created and filled, False for any error
+ */
+static gboolean
+gst_tensor_src_iio_create_config (GstTensorSrcIIO * tensor_src_iio)
+{
+  GList *list;
+  GstTensorSrcIIOChannelProperties *channel_prop;
+  gint tensor_info_merged_size;
+  gint info_idx = 0, dim_idx = 0;
+  /**
+   * create a bigger array, insert info in it and
+   * then merge tensors with same type+size
+   */
+  GstTensorInfo info[tensor_src_iio->num_channels_enabled];
+
+  /** compile tensor info data */
+  for (list = tensor_src_iio->channels; list != NULL; list = list->next) {
+    channel_prop = (GstTensorSrcIIOChannelProperties *) list;
+    if (!channel_prop->enabled)
+      continue;
+    info[info_idx].name = channel_prop->name;
+    info[info_idx].type =
+        gst_tensor_src_iio_get_type_for_channel (channel_prop);
+    for (dim_idx = 0; dim_idx < NNS_TENSOR_RANK_LIMIT; dim_idx++) {
+      info[info_idx].dimension[dim_idx] = 1;
+    }
+    info_idx += 1;
+  }
+  g_assert_cmpint (info_idx, ==, tensor_src_iio->num_channels_enabled);
+
+  /** merge info about the tensors with same type */
+  tensor_info_merged_size = tensor_src_iio->num_channels_enabled;
+  if (tensor_src_iio->merge_channels_data) {
+    tensor_info_merged_size =
+        gst_tensor_src_merge_tensor_by_type (info,
+        tensor_src_iio->num_channels_enabled);
+  }
+
+  /** verify the merging of the array */
+  if (tensor_info_merged_size < 0) {
+    GST_ERROR_OBJECT (tensor_src_iio, "Mismatch while merging tensor");
+    goto error_ret;
+  } else if (tensor_info_merged_size == 0) {
+    GST_ERROR_OBJECT (tensor_src_iio, "No info to be merged");
+    goto error_ret;
+  } else if (tensor_info_merged_size > NNS_TENSOR_SIZE_LIMIT) {
+    GST_ERROR_OBJECT (tensor_src_iio,
+        "Number of tensors required %u for data exceed the max limit",
+        tensor_info_merged_size);
+    goto error_ret;
+  }
+
+  /** tensor/s config data */
+  if (tensor_info_merged_size == 1) {
```

Comment:
[trivial suggestion] use only tensors config (self->config)
1. remove GstTensorSrcIIODataConfig and set GstTensorsConfig in data structure
2. to get caps from config
if is_tensor
    GstTensorConfig c;
    gst_tensor_info_copy (&c.info, &self->config.info.info[0]);
    c.rate_n = self->config.rate_n;
    c.rate_d = self->config.rate_d;
    caps = gst_tensor_caps_from_config (&c);
else
    caps = gst_tensors_caps_from_config (&self->config);

Reply:
This way would create an array for 16 (in other/tensors) but never use more than 1 element for other/tensor case. On the other hand, this method might make code simpler.
I am not sure which way is better - @myungjoo @again4you @leemgs - can you please comment on this?

Reply:
You may neglect the wasted memory specifically for that; using TensorsConfig means that we've decided to neglect that waste for the sake of code generality. :)

As long as we use GstBuffer (it has a static array of 16), we always waste that amount of memory. It makes all codes to use the same data structure, making it easier to read and to maintain.

Reply:
Updated, thanks for clarification.